src/connection.h

```diff
@@ -245,14 +359,81 @@ int connDisableTcpNoDelay(connection *conn);
 int connKeepAlive(connection *conn, int interval);
 int connSendTimeout(connection *conn, long long ms);
 int connRecvTimeout(connection *conn, long long ms);
-int connPeerToString(connection *conn, char *ip, size_t ip_len, int *port);
-int connFormatFdAddr(connection *conn, char *buf, size_t buf_len, int fd_to_str_type);
-int connSockName(connection *conn, char *ip, size_t ip_len, int *port);
-const char *connGetInfo(connection *conn, char *buf, size_t buf_len);
 
 /* Helpers for tls special considerations */
-sds connTLSGetPeerCert(connection *conn);
-int tlsHasPendingData();
-int tlsProcessPendingData();
+static inline void *connTypeGetCtx(ConnectionType *ct) {
+    return ct->get_ctx();
+}
+
+static inline void *connTypeGetClientCtx(ConnectionType *ct) {
+    return ct->get_client_ctx();
+}
```

Comment:
i'm trying to understand how generic these are and if they indeed need to be part of the interface.
i.e. if they can ever serve any other connection type, or just for extensions re-implementing the TLS one.
same about connTypeConfigure, and obviously connGetPeerCert.

one other alternative is to explicitly call a TLS specific function. i.e. the code that uses all of these is TLS specific code, doing something like `connTypeGetClientCtx(connectionTypeTls());` or `connTypeConfigure(connectionTypeTls(), &server.tls_ctx_config, 1)`.

@yossigo i'd like your feedback here.

Reply:
What about add a new API like `connCtrl()` style, and wrapper TLS specified function into this?

Reply:
@oranagra @pizhenwei I don't think exposing the OpenSSL context should be part of the interface, because we have no guarantee that any TLS implementation is going to have OpenSSL. If we believe it is required for certain operations, we should probably define a more abstract interface for those.

A generic `connGetPeerCert` is reasonable. We could insist on differentiating generic from TLS connection interfaces and only supporting it in the latter, but I'm not sure if it's worth the trouble - it can remain part of the basic interface and just be left out if not applicable.